Allow running console, server, dbconsole and runner from Engine

cllns opened the pull request:
Right now, if you're in the root directory of a Rails app, you can run `rails [command]`, where command is any of the supported commands (`generate`, `destroy`, `console`, `server`, `dbconsole`, `runner`, `test`, `plugin new`). You can also run any of those in most subdirectories (`app/models/`, `config/initializers/`, etc). 

The only subdirectories you **cannot** run all those commands in, is within an Engine. When you're in an engine folder, you only get access to `generate`, `destroy` and `test`. 

This patch fixes that, so if you run, e.g. `rails console` from within an engine, a console will be opened in the parent application. The way I did this is by traversing up the directory structure until we find a `config/application.rb` file (since Engine's seem to, by design, not have access to the application they're mounted in). 

I'm working on an app that has several engines, and it's a daily annoyance that I try to run `rails console` or `rails server` from an Engine folder, and it doesn't work. Then I have to `cd ..` and try again.

A few things:
- How to test this? I can't find any tests that make sure the command `rails console` works in normal apps. Just that we can call `Rails::Console.new(...)`.
- How this will work for people developing 'standalone' Engines that are gems (like Gemified assets)? Running these newly supported commands there should either not work, or try to run from their `dummy_app` instead. If we do the latter, which environment should we pick for Engines within a Rails app that also have a dummy app? The main application or the dummy app?
- I assume the engine is a subdirectory of the main app. If it's not, these commands won't work, but we should display a good error message. 
- I assume we're working with mountable engines, since that's what I have the most experience with. How will this work for other engines/plugins? I think fine, but this should be investigated.
- `help` right now shows a normal application HELP_MESSAGE, not the one for engines. This is an easy fix, but it'd make the diff much busier. 
- I don't have support for 'plugin new', since that seems like something you probably should do from the main app.
- I'd like to make this functionality into a class, which inherits from the `CommandsTask` class. Then we can just leverage `railties/lib/rails/commands.rb` instead of having this special file for Engines. If I get the go-ahead, I'll work on that.

Thanks for reading. Any thoughts?

cllns commented:
Looks like the travis build failure is unrelated to my code: `ActiveModel::Type::IntegerTest#test_casting_nan_and_infinity`

cllns commented:
Any thoughts @schneems? It'd be cool to get this in to `5.0` if possible :)

schneems commented on railties/lib/rails/engine/commands.rb:
```diff
@@ -19,6 +23,20 @@
+when 'console', 'server', 'dbconsole', 'runner', 'version', 'help'
+  require "rails/commands/commands_tasks"
+
+  while Dir.pwd != "/"
```
This will never be true on windows. Is Rails.root set and valid here?

cllns replied:
Good call. I just switched it to `Pathname.pwd.root?`, which says it has 'experimental' support for non-Unix platforms. Rails.root is not net unfortunately.

schneems commented:
What do you think @rafaelfranca

rafaelfranca commented:
:shipit:

arthurnn commented:
@cllns Thanks for the PR. Do you mind squashing your commits in one?

Also we will need a changelog for this. Thanks

cllns commented:
Will do. The merge-conflict is pretty rough, because the functionality from the file was extracted into a class, so I'll have to take care of that.

cllns committed:
Refactor Rails::Engine::CommandsTasks to inherit from Rails::CommandsTasks

cllns commented:
Hey @arthurnn, I just refactored the code to work with the new `Rails::Engine::CommandsTask` class. That class seemed to have been copied and modified from `Rails::CommandsTasks`, so I refactored it to inherit from that class. It makes the code a lot cleaner, I think. Net negative 28 lines :smile: 

In the process, I also made some other minor modifications, like adding a `MAIN_COMMANDS` constant, rather than just having it be in the text. I made that as a hash, instead of a nested array, and converted `ADDITIONAL_COMMANDS` to be a hash as well. This makes it easier for me to exclude some commands in the subclass.

The code currently works for all the non-Rake commands, but I'm not sure how to proceed for the commands that proxy to rake. Any advice? Some, like `rails db:migrate` should migrate within the engine's test app, not the application, but others like `rails routes` should run in the application.

Should I just go through and figure out which rake tasks should be run where, and add them to `APPLICATION_WHITELIST` and `ENGINE_WHITELIST`? 

Note that I've manually tested this functionality, but there aren't any real (i.e. code) tests. Also, I still need to test this for standalone Engines, since they're not a part of a Rails app, but may have a `dummy_app`.

arthurnn commented:
Sorry @cllns, took me a while to come back to this ,
Do you mind re-working on this, making sure there is no conflict and CI is green?

Thanks a bunch

kaspth commented:
We've got a new command infrastructure set up now and every command should now work in engines as well.

And they do work if I generate an engine and run this from test/dummy:

```
bin/rails runner 'puts Rails.env'
config.eager_load is set to nil. Please update your config/environments/*.rb files accordingly:

  * development - set it to false
  * test - set it to false (unless you use a tool that preloads your test environment)
  * production - set it to true

development
```

However, it doesn't seem to execute `environments/development.rb` (I added a `raise` in there and it still went off fine).

When generating the engine with `/path/to/local/rails/railties/exe/rails plugin new pluginer --engine`, the `bin` directory had the whole local Rails path prefixed.

@y-yagi you've been helping me out with engine things before, are you interested in fixing those two issues above?

Thanks, @cllns!

cllns commented:
Great! Is this only on `master`? Because I have an app with Engines, that I generated using `5.0.0.1` and I can't run `rails console` (or server) from within the engine's folder

kaspth commented:
Yep, only on master.

y-yagi commented:
> are you interested in fixing those two issues above?

Sure!  I'll do.

cllns commented:
Thanks folks :D

kaspth commented:
Awesome! Thanks :)
